specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+Index data, once read from the CARv2 container, provides a mapping of hash digest bytes to block location in byte offset from the beginning of the CARv2 container. The index only uses the hash digest&mdash;it does not use the full bytes of a CID, nor does it use any of the multihash prefix bytes.
+
+The first byte(s) of a CARv2 index are a varint that indicates the index format type. The remaining bytes follow the encoding rules of that index format type.
+
+As the index only contains the hash digest bytes, other details contained within the block's CID and the length of the block's bytes must be derived by inspecting the initial bytes of the block entry within the data payload.
+
+**TODO:** How are we handling identity hash CIDs?
+* Do we want to state that identity multihash CIDs are explicitly excluded from indexes? 
+* Can we defer an assumption that a block store interface would return the bytes without having to look them up when it sees an identity CID? 
+* What are the implications of having an index that doesn't contain all of the blocks within the CAR?
+* Or do we want to explicitly exclude index CIDs entirely from the CARv2 format since they are arguably a waste of space?
+
+#### Format `0x300000`: IndexSorted
+
+**TODO:** I'm assuming this is supposed to be a multicodec and that `0x300000` is using the reserved range - so we should register this in the table.
+
+An index format type of `0x300000` indicates the "IndexSorted" format.
+
+IndexSorted sorts hash digests by two dimensions: first into buckets of _digest length_, smallest to largest, and then within those buckets ordered by a simple byte-wise sorting. In this way, locating a hash digest within the CAR requires first finding the bucket matching the length of the requested hash digest, then searching the ordered list of digests within that bucket to find the matching entry.
+
+* IndexSorted may contain one or more length-grouped buckets of digests.
+* Buckets are ordered by digest length and concatenated together to form the index.
+* Each bucket is prefixed with:
+  * a "width" encoded as a 32-bit **signed** little-endian integer indicating the common byte length of the combination of hash digests and their offsets in this bucket; followed by
+  * a "count" encoded as a 64-bit **signed** little-endian integer which determines the total number of hash digests (and their offsets) bucket.
+
+A common case of a single bucket of 32-byte hash digests is expected due to the commonality of this digest length for CIDs.
+
+Individual index entries are the concatenation of the hash digest an an additional 64-bit **unsigned** little-endian integer indicating the offset of the block from the begining of the CARv2 container. Offsets locate the first byte of the varint that prefix the `CID:Bytes` pair within the CARv1 payload. See the [data section in the CARv1 Specification](../carv1/index.md#data) for details on block encoding.
```

Comment:
Thinking to myself for a second: if the indexed offsets are from the beginning of the CARv2 container, this means that changing the Data Offset (padding before CARv1) requires changing all the offsets in the index. I guess that won't be a common operation, but another option would be to make all these offsets relative to Data Offset, i.e. relative to the start of the CARv1.

Reply:
Thanks to @masih for pointing out that the intention was to be offsets from the beginning of the CARv1, so this is wrong. It ought to be possible to hold the index and CARv1 in separate files and for them still to make sense without modification.